src/persist-client/src/internal/machine.rs

```diff
@@ -159,14 +177,12 @@ where
             })
             .await;
 
-        if !self.applier.state().collections.is_tombstone()
-            && !self
-                .applier
-                .state()
-                .collections
-                .leased_readers
-                .contains_key(reader_id)
-        {
+        let reader_already_expired = self.applier.read_locked_state(|state| {
```

Comment:
i think this introduces a race where the reader does get registered and then expired before we get this lock. I think we could either (1) remove these checks entirely or (2) run the expiry logic _before_ the cmd closure in the apply loop